Refactor improvements sebastian

jnaulty opened the pull request:
hey @sebastianriese 
I thought these might be some minor improvements on your refactor PR. What do you think?
- add requirements.txt (makes it easier for setting up `virtualenvironments` (pretty much just 
```
#> python3 -m venv ../openeit_venv; 
#> ../openeit_venv/bin/activate; 
#> pip3 install -r requirements.txt
```
- move module to lowercase module name `openeit` [PEP-8](https://www.python.org/dev/peps/pep-0008/#package-and-module-names)
- add an `app.py` file to import and run from (makes importing package.submodules a whole lot easier imo).
- run dashboard with `python3 run.py` from root directory
- change `setup.py` to not be based off of `py2app` It has a more generic structure with some basic info. I believe it's enough info to publish to pypi, if one were inclined to do that. This allows for the usual `python3 setup.py install/deploy/build/bdist_rpm` commands :smile:

jnaulty committed:
move OpenEIT to openeit

jnaulty committed:
- can run now with `python3 run.py`
- moved logic to main() function for starting gui

jnaulty committed:
add requirements.txt

- can now install with `pip install -r requirements.txt`

jnaulty committed:
change setup.py to point to run.py

jnaulty committed:
move app-logging to global scope

jnaulty committed:
find_packages for setup.py

sebastianriese commented on openeit/app.py:
```diff
@@ -8,7 +8,7 @@
 import serial.tools.list_ports
 
-import OpenEIT.dashboard
+import openeit.dashboard
```
I actually think OpenEIT is the better name than (openeit) (but am willing to change it, if the community at large prefers this). While this is in violation of PEP8 it is a conscious violation of PEP8 (which PEP8 explicitly supports).

My reasoning is that (a) it is in concordance with many other packages, e.g. OpenSSL, OpenGL, etc. of the form OpenTLA (b) openeit is hard to read and underscores are discouraged in package names.

sebastianriese commented on requirements.txt:
```diff
@@ -0,0 +1,15 @@
+cycler==0.10.0
```
The depenencies given by requirements.txt are just plain wrong! The code does not use most of the packages listed.

JeanRintoul replied:
I'd like to use a package handler such as anaconda which is used broadly in many companies to avoid module conflicts.

jnaulty replied:
@sebastianriese 
Actually, these packages are required for running the dashboard.
`cycler` dependency is pulled in from `matplotlib`
```
Requirement already satisfied: cycler>=0.10 in ./myvenv/lib/python3.4/site-packages (from matplotlib>=1.3.1->scikit-image)
```
I'm pretty sure this is the minimal set of required packages to get the dashboard up and running. But it's easy to check, just setup a new virtualenv and try running the dashboard.

@JeanRintoul :
As far as an anaconda based package handling, I think a `requirements.txt` file can help bootstrap a project. And it's incredibly useful (and basically required) for anyone _not_ using anaconda.

sebastianriese replied:
I fully agree we should also support people who do not use anaconda (for example using anaconda would be an inconvenience for me, most of my python packages simply come from my Linux distribution, to use anaconda I would have to install things twice and switch between Python environments). But I think we should only add *immediate* dependencies to our requirements, that is those packages that we import directly. Adding transitive dependencies is a can of worms we do not want to open! We simply cannot keep up with the dependency changes of our dependencies; that's what package managers are for, pip should be quite able to handle that by itself, as are the package managers of any decent distribution. Also, I am strongly against requiring specific versions with `==`, this is overly specific (and will cause redundant installations of common packages), we should use `~=` (compatible version) unless there is known breakage (or that version is antique). In short: requirements.txt should only contain the packages that we use directly and should not prescribe specific versions, doing so will deter people from installing and using the software.

katrinabrock replied:
>we should only add immediate dependencies to our requirements

+1 If an immediate dependency drops one of *its* dependencies, there's no reason for users to install what is now essentially a random package.

>I am strongly against requiring specific versions with `==`

+1 If anyone wants to install this package with the same env as other packages, and another package with the same dependency, you can create unnecessary and difficult to resolve version conflicts by being overly specific on versioning.

sebastianriese commented on setup.py:
```diff
@@ -1,19 +1,33 @@
+VERSION = '0.1.0'
+AUTHOR = 'Jean Rintoul'
+LICENSE = 'LICENSE'
+INSTALL_REQUIRES = [
```
Same problem as in requirements.txt

jnaulty replied:
responded to `requirements.txt` problem above.

sebastianriese commented on setup.py:
```diff
@@ -1,19 +1,33 @@
+    version=VERSION,
+    author=AUTHOR,
+    license=LICENSE,
+    packages = find_packages(),
```
Have you tested that find_packages only finds the openeit package (there is a lot of cruft in the repo).

JeanRintoul replied:
yeah, there is a whole folder of semi-useful cruft that could later be deleted(it's useful for development of new algorithms however but not formalized in any way).

sebastianriese replied:
Yes, I agree that it is useful to keep it for the moment, since it contains extra testing data and useful "cookbook" like parts for further work.

jnaulty replied:
This type of problem can be addressed with passing kwargs to `find_packages(where=, exclude=)`.

sebastianriese commented on openeit/app.py:
```diff
@@ -8,7 +8,7 @@
 
```
I disagree with adding the actual app to the package (but again, this is a personal preference). My argument is, that the package is a framework, that is independent of the presentation, it supplies infrastructure for image reconstruction, connecting to hardware devices and coordinating those (openeit.controller which is admittedly perhaps already to specific), but it does *not* depend on a GUI toolkit or specific UI.

In this sense dashboard.py is an application using the framework openeit, I plan to build dashboard_qt.py that reimplements the dashboard using Qt, I would not want to have that code hanging in the OpenEIT package – both dashboard front-ends could be supported by the same backend OpenEIT!

jnaulty committed:
Move openeit back to OpenEIT

In agreement with @sebastianriese:
```
While this is in violation of PEP8 it is a conscious violation of PEP8 (which PEP8 explicitly supports).

My reasoning is that (a) it is in concordance with many other packages, e.g. OpenSSL, OpenGL, etc. of the form OpenTLA (b) openeit is hard to read and underscores are discouraged in package names.
```

jnaulty committed:
change requirements.txt and setup.py to have compatible release version specifier

jnaulty committed:
move OpenEIT/app.py back to dashboard.py
This should probably live in an 'examples/' directory.
OpenEIT is providing the necessary framework for such apps/dashboards to exist.
But this framework should be somewhat independent of the implemented dashboard.

jnaulty commented:
Hey @sebastianriese,

Wanted to get back to you sooner w.r.t. your comments and input.
Thanks for the feedback. I hope to have addressed some of your main concerns:
- `dashboard.py --> openeit/app.py` is no longer a thing (I like the direction of keeping this a somewhat independent framework).
- the `requirements.txt` + `setup.py` have compatible versions and dropped secondary packages pulled in by the main required packages.
- and `openeit` is now renamed back to `OpenEIT`

Also, thanks @katrinabrock for reviewing this and giving comments too :) :robot:
